packages/framework-provider-aws/src/library/events-adapter.ts

```diff
@@ -101,28 +103,78 @@ export async function storeEvents(
+     * Entity snapshots can be potentially created by competing processes, so we need to make sure
+     * that no matter what snapshot we find, we can always rebuild the correct state by
+     * replaying all the events that happened after the snapshot we take as the origin.
+     * This way of storing the data makes snapshot caching an idempotent operation, and allows us to
+     * aggressively caching snapshots without worrying about insertion order.
+     */
+    const sortKey = snapshotEnvelope.snapshottedEventPersistedAt
+    await dynamoDB
+      .put({
+        TableName: config.resourceNames.eventsStore,
+        ConditionExpression: `${eventsStoreAttributes.partitionKey} <> :partitionKey AND ${eventsStoreAttributes.sortKey} <> :sortKey`,
+        ExpressionAttributeValues: {
+          ':partitionKey': partitionKey,
+          ':sortKey': sortKey,
+        },
+        Item: {
+          ...snapshotEnvelope,
+          [eventsStoreAttributes.partitionKey]: partitionKey,
+          [eventsStoreAttributes.sortKey]: sortKey,
+          [eventsStoreAttributes.indexByEntity.partitionKey]: partitionKeyForIndexByEntity(
+            snapshotEnvelope.entityTypeName,
+            snapshotEnvelope.kind
+          ),
+        },
+      })
+      .promise()
+  } catch (e) {
+    const error = e as Error
+    if (error.name == 'ConditionalCheckFailedException') {
+      throw new OptimisticConcurrencyUnexpectedVersionError(error.message)
+    }
+    throw e
+  }
+}
+
 async function persistEvent(
   dynamoDB: DynamoDB.DocumentClient,
   config: BoosterConfig,
-  eventEnvelope: EventEnvelope
+  eventEnvelope: NonPersistedEventEnvelope
 ): Promise<void> {
   try {
-    const partitionKey = partitionKeyForEvent(eventEnvelope.entityTypeName, eventEnvelope.entityID, eventEnvelope.kind)
+    const partitionKey = partitionKeyForEvent(eventEnvelope.entityTypeName, eventEnvelope.entityID)
     // Generate a new timestamp as sorting key on every try until we can persist the event.
     // This way we guarantee ordering even with events that are stored in the same millisecond
-    const sortKey = new Date().toISOString()
+    const persistableEvent: EventEnvelope = {
+      ...eventEnvelope,
+      persistedAt: new Date().toISOString(),
+    }
     await dynamoDB
       .put({
         TableName: config.resourceNames.eventsStore,
         ConditionExpression: `${eventsStoreAttributes.partitionKey} <> :partitionKey AND ${eventsStoreAttributes.sortKey} <> :sortKey`,
         ExpressionAttributeValues: {
           ':partitionKey': partitionKey,
-          ':sortKey': sortKey,
+          ':sortKey': persistableEvent.persistedAt,
         },
         Item: {
           ...eventEnvelope,
           [eventsStoreAttributes.partitionKey]: partitionKey,
-          [eventsStoreAttributes.sortKey]: sortKey,
+          [eventsStoreAttributes.sortKey]: persistableEvent.persistedAt,
```

Comment:
Same here: Doing this we're overriding the `createdAt` that comes in the `eventEnvelope`by the calculated `persistedAt` value (see line 164 in this same file)